tests/Microsoft.Identity.Test.Unit/CacheTests/CacheSyncronizationTests.cs

```diff
@@ -0,0 +1,123 @@
+                                                              .WithCacheSynchronization(useCacheSyncronization)
+                                                              .WithHttpManager(httpManager)
+                                                              .BuildConcrete();
+
+                httpManager.AddMockHandlerSuccessfulClientCredentialTokenResponseMessage();
+
+                BlockingCache inMemoryTokenCache = new BlockingCache();
+                inMemoryTokenCache.Bind(app.AppTokenCache);
+
+                // Seed the cache with a token
+                var result = await app.AcquireTokenForClient(TestConstants.s_scope.ToArray())
+                    .ExecuteAsync()
+                    .ConfigureAwait(false);
+                Assert.IsTrue(result.AuthenticationResultMetadata.TokenSource == TokenSource.IdentityProvider);
+
+
+                var blockingTask = RunAsync(inMemoryTokenCache, app, true);
+                var nonBlockingTask1 = RunAsync(inMemoryTokenCache, app, false);
+                var nonBlockingTask2 = RunAsync(inMemoryTokenCache, app, false);
+
+                int res = Task.WaitAny(new[] { blockingTask, nonBlockingTask1, nonBlockingTask2 }, 100);
+
+
+                if (useCacheSyncronization)
+                {
+                    Assert.AreEqual(-1, res, "WaitAny should have timed out, all tasks are blocked when the first call is blocking");
+                }
+                else
+                {
+                    Assert.AreNotEqual(-1, res, "WaitAny should have NOT timed out, the 2 non-blocking tasks should be allowed to complete");
+                    Assert.IsTrue(nonBlockingTask1.IsCompleted);
+                    Assert.IsTrue(nonBlockingTask2.IsCompleted);
+                    Assert.IsFalse(blockingTask.IsCompleted, "The blocking task should still be blocked");
+                    Assert.IsTrue(nonBlockingTask1.Result.AuthenticationResultMetadata.TokenSource == TokenSource.Cache);
+                    Assert.IsTrue(nonBlockingTask2.Result.AuthenticationResultMetadata.TokenSource == TokenSource.Cache);
+                }
+            }
+        }
+
+        private static async Task<AuthenticationResult> RunAsync(BlockingCache cache, IConfidentialClientApplication app, bool block = false)
+        {
+            cache.BlockAccess = block;
+            return await app.AcquireTokenForClient(TestConstants.s_scope.ToArray()).ExecuteAsync().ConfigureAwait(false);
+        }
+
+        public class BlockingCache
+        {
+            private byte[] _cacheData;
+
+
+            public bool BlockAccess { get; set; } = false;
+
+            public async Task BeforeAccessNotificationAsync(TokenCacheNotificationArgs args)
+            {
+                if (BlockAccess)
+                {
+                    // completely blocked
+                    await Task.Delay(10000).ConfigureAwait(false);
```

Comment:
Interesting! so that's what can happen with a Redis cache that loses its connection, being blocked for 15 mins

Reply:
Yes!